src/full/Agda/TypeChecking/Serialise/Base.hs

```diff
@@ -431,18 +431,20 @@ class VALU t b where
                Proxy t -> Node -> Maybe (Products (Constant Int32 (Domains t)))
 
 instance VALU t 'True where
-
+  {-# INLINE valuN' #-}
```

Comment:
Wow, great!
The obvious question, just to check:  Did you benchmark with Agda built with `-f+optimise-heavily`?
If yes, then apparently this flag isn't powerful enough to perform such optimizations...

Reply:
Oh, good shout, I always forget to rebuild with those flags enabled!

The improvements are a little smaller - 20% reduction in allocations, down from 25% - but still present.

There is a bit of me which is surprised by how much of an impact the change makes. My understanding is that GHC is normally pretty eager to inline functions which are immediately pattern matched on - and I'd expect `-fexpose-all-unfoldings` to only help with that - but clearly not!